### ISSUE TITLE
SSL Doesn't Work with AWS

### ISSUE THREAD

**Issue by danastasio** [limads/queries#1]
Hello!

First, I love this app. I've been waiting a long time for a gtk4 sql client, and I'm happy to see you're working on this. Thank you!

I can't connect to my remote postgres instance. It's hosted on an AWS server, and so the hostname doesn't match what I put in the configuration fields. Looks like you're using sslmode=verify-full, which simply won't allow this connection. If you allow us to use verify-ca instead then it will work if I provide the pem file for AWS.

Error from psql:

psql: error: connection to server at "<setting hostname>" (<ip address>), port 5432 failed: server certificate for "<aws hostname>" does not match host name "<setting hostname>"

Error from Queries:

Error performing TLS handshake: error: 0A0000SSL routines:tls_post_process_server_certificate:certificate verify failed:../ssl/statem/statem/clnt.c:1887:: hostname...

The error times out after like three seconds so I had to do a bunch of takes to get that right. Hope there are no typos.

Thanks! I'm excited to start using this.

**Comment by limads** [comment]
Hello, thanks for the feedback!

This error message means Queries tried to connect to your remote database via TLS instead of SSL. When you added your certificate, did you select the SSL mode in the toggle button to the right of the certificate path entry?

**Comment by danastasio** [comment]
> When you added your certificate, did you select the SSL mode in the toggle button to the right of the certificate path entry?

Yes, before opening this issue I tried selecting both TLS and SSL. I also tried using the IP address that the hostname resolves to,  and the AWS hostname, but they either don't work or timeout. I can connect to the db with psql, but I need to use verify-ca, not verify-full.

**Comment by danastasio** [comment]
I was going to submit a pull request with what I think could fix this, but it won't build on my computer. To test I was going to put 

```rust
uri += "?sslmode=verify-ca";
```
in https://github.com/limads/queries/blob/master/src/client/conn.rs#L543

If that worked I was going to build out a button for a toggle. I think that's all that's needed to resolve this.

**Comment by limads** [comment]
Hello danastasio,

After investigating this issue, I think your solution wouldn't work since `tokio_postgres` does not currently support the verify-ca or verify-full connection string options (see [this](https://github.com/sfackler/rust-postgres/issues/768) issue). I've tried those options and the driver refuses any connections while they are set. The application will be using `sslmode=require` for the time being, but the option could certainly be added once the driver supports them.

Your error message explicitly mentions a TLS handshake, which shouldn't happen if you had SSL connection selected, so I suspect a bug with the certificate matching could be causing this. I don't think this is anything AWS-specific, since I can connect to a remote AWS database myself.

I've just pushed an update to the application which does impact certificate matching, can you please verify if the problem persists?

```
flatpak update io.github.limads.Queries
```

**Comment by danastasio** [comment]
Hello,

Yes the problem persists. The error message is slightly different. It now ends in: certificate_verify_fail. The TLS handshake part remains. I tried TLS and SSL. It does work using psql and sslmode=require.

I've also tried using the .pem file directly from AWS as well as ~/.postgresql/root.crt (which I have as a symlink to /etc/ssl/certs/ca-certificates.crt)

Thanks!

**Comment by limads** [comment]
I think it is possible to configure the TLS connectors handed to `tokio_postgres` in such a way that they parallel the semantics of `psql` ('verify-ca' and 'verify-full'), even if the driver does not expose such configurations. I have been using the default connectors successfully myself, but if this turns out to be a common issue, the application should have the options exposed in the security settings GUI. Give me a few days to study this possibility.

**Comment by limads** [comment]
Are you by any chance using AWS's `global-bundle.pem`? I've had problems with this one (not exactly the problem your reported) but you could try the regional bundle instead.

Looking at the AWS ![docs](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/PostgreSQL.Concepts.General.SSL.html), `verify-full` should be supported:

> When you connect using SSL, your client can choose whether to verify the certificate chain. If your connection parameters 
> specify sslmode=verify-ca or sslmode=verify-full, then your client requires the RDS CA certificates to be in their trust store or 
> referenced in the connection URL. This requirement is to verify the certificate chain that signs your database certificate.

Perhaps this could be related to some database configuration set at your instance?

I've just pushed another update to the application, please update before you try a new connection.

**Comment by danastasio** [comment]
Yes, I am using the global-bundle.pem. I'm actually not sure what region the DB is hosted in (we are end users, not admins). I'll try some of the regional .pem files on Monday and see if that helps.

Thanks!

**Comment by danastasio** [comment]
Ok. I was able to get it working. I had to disable hostname verification since I connect to db.company.com but that resolves to ecX.ip-address.compute-X.amazonaws.com. Once I did that it worked a treat.

I'd like to submit a pull request for this, but since I'm not a GUI person (and so need to learn how to do so) I'd like to know if this is a pull request you're willing to accept? I just don't want to put work into a pull that will ultimately not be merged.

Thanks!

**Comment by limads** [comment]
Hi @danastasio, thanks for looking into this, I'm glad you could connect to your database successfully. 

I could add a "Enable server name identification" switch at the security settings, perhaps for each host separately. My main concerns are:

- Users without clear knowledge about the consequences of disabling host name identification might switch it off while just trying to make things work, therefore exposing them to security vulnerabilities. 
- I would like the GUI to stay agnostic to database engine, and I don't know yet if all drivers Queries could eventually support (MySQL, MariaDB, etc) will all support enabling/disabling host name identification.

This could also be done via environment variables (e.g. QUERIES_DISABLE_SNI that users could configure at Flatseal or something like that), thus keeping it away from the main application UI design and treating it more like a advanced feature.

**Comment by limads** [comment]
Hello @danastasio, after some thought I have decided to add the option of disabling hostname verification to the UI. Verification is enabled by default, so users have to actively disable it for each connection separately (hopefully the label below the switch is enough to dissuade liberal use of the feature). The other big future candidate database driver (MySQL) also works via `native_tls`, so I think it is safe to assume the UI will stay reasonably generic across different drivers.

Can you please update and verify if you can establish your connection?

`flatpak update io.github.limads.Queries`
